🏗️ Build spider: Spokane Valley City Council

SimmonsRitchie opened the pull request:
## What's this PR do?
Adds a spider to scrape Spokane Valley City Council meetings.

## Why are we doing this?
Requested by City Bureau's Spokane partners.

## Steps to manually test

1. Ensure the project is installed:
```
pipenv sync --dev
```
2. Activate the virtual env and enter the pipenv shell:
```
pipenv shell
```
3. Run the spider:
```
scrapy crawl sv_city_council -O test_output.csv
```
4. Monitor the output and ensure no errors are raised.

5. Inspect `test_output.csv` to ensure the data looks valid. Target website is [here](https://spokanevalley.granicus.com/ViewPublisher.php?view_id=3).

## Are there any smells or added technical debt to note?
- This spider intentionally ignores the target's `robots.txt` file in order to parse meeting data. That is not preferable but there appears to be a precedent within the city-scrapers project of doing this. Eg. [this spider](https://github.com/City-Bureau/city-scrapers/blob/f897d5191396ced70a84f1be0630c3e6d81a25fc/city_scrapers/spiders/il_elections.py) in `city-scrapers`. Many of its sister spiders do the same.
- This page is a bit lacking in detailed meeting information, but we're parsing what we can. Helpfully, the page includes unix timestamps for meeting start times. However, this required some nuanced handling in my `parse_start` method because city-scrapers-core doesn't expect datetimes that are timezone aware.

SimmonsRitchie committed:
Build sv_city_council spider

SimmonsRitchie committed:
Fix datetime handling

We need to ensure the meeting start time is converted into a timezone naive date in Pacific time.

SimmonsRitchie committed:
Add tests for sv_city_council

LienDang commented:
Hi @SimmonsRitchie There are some naming conventions to develop the `city-scrapers-*` repo that we need to follow, so I'd like mention here (I think we should document them somewhere):
- On documenters.org side, while setting up a new city, we'll create a list of Azure variables that we'll pass to the city-scrapers team to set up in the section Github Actions secrets for `city-scrapers-*` repo, eg. for this repo https://github.com/City-Bureau/city-scrapers-spokane the Azure variables were set up [here](https://github.com/City-Bureau/city-scrapers-spokane/settings/secrets/actions). 

<img width="610" alt="Azure variables for city-scrapers-spokane repo" src="https://github.com/City-Bureau/city-scrapers-spokane/assets/1421838/2af01207-c367-4cea-8589-bd016624cac8">

- In the variables list, you'll see the variable `AZURE_CONTAINER=meetings-feed-spo`, it means:
  - the city-scrapers devs need to create a container `meetings-feed-spo` with Azure storage account `cityscrapers` (which was done by someone (I think @haileyhoyat did that) at 9/18/2023, 11:58:52 PM)
  - All the spiders file within folder `city-scrapers/spiders/` must start with the prefix `spo_` (the last 3 characters in the Azure container name and the underscore `_`). So in this PR, the file `city_scrapers/spiders/sv_city_council.py` must be renamed to `spo_*.py`

LienDang commented on tests/test_sv_city_council.py:
```diff
@@ -0,0 +1,89 @@
+def test_links():
+    assert parsed_items[0]["links"] == [
+        {
+            "href": "//spokanevalley.granicus.com/AgendaViewer.php?view_id=3&event_id=1314",  # noqa
```
@SimmonsRitchie can you please prepend the protocol `https` to this `href` value in the `_parse_links` function? So it is expected as
`"href": "https://spokanevalley.granicus.com/AgendaViewer.php?view_id=3&event_id=1314"`

SimmonsRitchie replied:
Thank you for catching that, @LienDang! Fix made. Added a method to ensure the output URLs always include the `https` protocol.

SimmonsRitchie committed:
Prefix spider with "spo_"

The "spo_" prefix is needed for this spider based on the Azure configuration for this repo.

SimmonsRitchie committed:
Add guard against bad meeting data

SimmonsRitchie committed:
Ensure links contain https protocol

SimmonsRitchie commented:
Thank you for the review, @LienDang! That is very helpful background, re: Azure containers. Those particular env vars have been a bit mysterious to me until now. Changes:

- I have now prefixed the spider, testing suite, and related references with `spo_`.
- Per your [comment](https://github.com/City-Bureau/city-scrapers-spokane/pull/12#discussion_r1474088386), I've improved the URL output for parsed links.
- I discovered today that sometimes a row contains invalid data if a meeting is "In Progress", I added an extra check in the spider to avoid errors if this happens.

LienDang commented on city_scrapers/spiders/spo_sv_city_council.py:
```diff
@@ -99,8 +100,15 @@ def _parse_links(self, item):
+        """Ensure the URL starts with 'https://', fixing incomplete
+        or incorrect parts."""
+        pattern = r"^(ttps://|tps://|ps://|s://|://|//|/)"
+        return re.sub(pattern, "https://", url, count=1)
```
Hi @SimmonsRitchie Not sure if you should capture many patterns here. IMO I would handle 2 cases:
- If the url starts with 2 slashes`//` .eg in this case `//spokanevalley.granicus.com/AgendaViewer.php?view_id=3&event_id=1314` I will prepend it with the protocol of the current browsing site, so it could be `http` or `https`. In this case, the `start_url` is https://spokanevalley.granicus.com/ViewPublisher.php?view_id=3 so I will extract the protocol `https` from the `start_url` and prepend it to this `url`
- If the url starts with single slash `/`, it means relative path, so I'll prepend the url with the protocol and domain name of `start_url`

I think there are some common Python libs to parse the url. Eg. in documenters project we use `urllib.parse.urlparse` to extract the protocol, domain and relative path,... I'd suggest you give it a try

SimmonsRitchie replied:
No problem, @LienDang ! I wasn't quite sure how thorough you were thinking in terms of pattern checking. I appreciate the extra detail you've provided. I've reworked the method to handle those two cases and added some tests for it. Let me know if that's okay or if you'd like additional changes 🔧

LienDang replied:
Hi @SimmonsRitchie I don't see your changed as you said `I've reworked the method to handle those two cases and added some tests for it`. Have you updated it in another PR? It's ok to merge then.

SimmonsRitchie replied:
Hah, that was dumb of me. Sorry, @LienDang, it looks like I forgot to push my changes. The remote branch is updated now.

SimmonsRitchie committed:
Rework URL cleaning func

SimmonsRitchie committed:
Add tests for URL cleaner func
